qemu/tests/thin_provisioning.py

```diff
@@ -0,0 +1,176 @@
+        scsi_disk_info = session.cmd_output('lsscsi').splitlines()
+        scsi_debug = [_ for _ in scsi_disk_info][-1]
+    scsi_debug = scsi_debug.split()
+    scsi_id = scsi_debug[0][1:-1]
+    device_name = scsi_debug[-1]
+    return scsi_id, device_name
+
+
+def get_provisioning_mode(pci_id, device, session=None):
+    """
+    Get disk provisioning_mode, value usually is 'writesame_16', depends
+    on params for scsi_debug module.
+
+    :param pci_id: Pci slot number
+    :param device: The disk name
+    :param session: Guest session
+    :return:
+    """
+    device_name = os.path.basename(device)
+    path = "/sys/block/{:s}/device/scsi_disk".format(device_name)
+    path += "/{:s}/provisioning_mode".format(pci_id)
+    try:
+        if session is None:
+            o = utils.read_one_line(path).strip()
+            return o
+        else:
+            o = session.cmd_output("cat %s" % path)
+            return re.search(r"(unmap)", o).group(1)
+    except IOError:
+        logging.error("Can't get the provisioning_mode"
+                      ", output is: {}".format(o), logging.error)
+
+
+def get_allocation_bitmap():
+    """
+    Get block allocation bitmap
+    """
+    path = "/sys/bus/pseudo/drivers/scsi_debug/map"
+    try:
+        bitmap = utils.read_one_line(path).strip()
+        return bitmap
+    except IOError:
+        raise exceptions.TestError("block allocation bitmap doesn't exists")
+
+
+def get_guest_discard_disk(session):
+    """
+    Get disk without partitions in guest.
+
+    :param session: Guest session
+    :return: Guest disk name
+    """
+    list_disk_cmd = "ls /dev/[shv]d*|sed 's/[0-9]//p'|uniq -u"
+    disk = session.cmd_output(list_disk_cmd).splitlines()[0]
+    return disk
+
+
+@error_context.context_aware
+def run(test, params, env):
+
+    """
+      'thin-provisioning' functions test using scsi_debug:
+      1) load scsi_debug module (writesame mode:lbpws=1)or(unmap mode:lbpu=1)
+      2) boot guest with scsi_debug emulated disk as extra data disk
+      3) rewrite the disk with /dev/zero in guest
+      4) check block allocation bitmap in host
+      5) format the disk with ext4 or xfs (with discard support filesystem)
+      6) execute fstrim command for the mount point
+      7) check block allocation bitmap updated in host
+
+      :param test: QEMU test object
+      :param params: Dictionary with the test parameters
+      :param env: Dictionary with test environment.
+      """
+
+    error_context.context(
+        "TEST STEP 1: Load emulated scsi disk before test", logging.info)
+    os_dep.command("lsscsi")
+    host_id, disk_name = get_scsi_disk()
+    provisioning_mode = get_provisioning_mode(host_id, disk_name)
+    logging.info(
+        "Scsi pci id and plot is: {:s} and disk name "
+        "is: {:s}".format(host_id, disk_name))
+    logging.info("Current provisioning_mode = {:s}", provisioning_mode)
+
+    # Bitmap must be empty before test
+    if get_allocation_bitmap():
```

Comment:
step should:
1.  boot guest with scsi_debug emulated scsi_disk
2. mkfs on the scsi_disk in guest.
3. get bitmap in host
4. write file in guest on the disk
5. rm the file in guest
6. check bitmap on host  should same with step3.